File: mobile/scripts/generate-abdominal-thumbnails.py
#!/usr/bin/env python3
"""
Generate clean, static thumbnail images from abdominal exercise GIFs.
Extracts a representative frame from each GIF to create stable thumbnails.
"""

import os
import sys
from pathlib import Path
from PIL import Image, ImageSequence

def extract_best_frame(gif_path, output_path):
    """
    Extract a clean, representative frame from a GIF.
    Uses the middle frame as it typically shows the exercise clearly.
    """
    try:
        with Image.open(gif_path) as gif:
            frames = list(ImageSequence.Iterator(gif))
            
            # Get middle frame (most representative of the exercise)
            middle_index = len(frames) // 2
            frame = frames[middle_index].convert('RGB')
            
            # Save as high-quality JPEG
            frame.save(output_path, 'JPEG', quality=95, optimize=True)
            return True
            
    except Exception as e:
        print(f"Error processing {gif_path}: {e}")
        return False

def main():
    # Setup paths
    base_dir = Path('/mnt/c/Users/PW1234/.vscode/new finess app/mobile/assets')
    gif_dir = base_dir / 'exercise-gifs' / 'abdominals'
    thumb_dir = base_dir / 'exercise-thumbnails' / 'abdominals'
    
    # Ensure thumbnail directory exists
    thumb_dir.mkdir(parents=True, exist_ok=True)
    
    # Process each GIF
    gif_files = list(gif_dir.glob('*.gif'))
    
    if not gif_files:
        print("No GIF files found in", gif_dir)
        return
    
    print(f"Found {len(gif_files)} GIF files to process")
    
    success_count = 0
    for gif_file in gif_files:
        # Create corresponding thumbnail filename
        thumb_name = gif_file.stem + '.jpg'
        thumb_path = thumb_dir / thumb_name
        
        print(f"Processing: {gif_file.name} -> {thumb_name}")
        
        if extract_best_frame(gif_file, thumb_path):
            success_count += 1
            print(f"  ✓ Created: {thumb_name}")
        else:
            print(f"  ✗ Failed: {thumb_name}")
    
    print(f"\nCompleted: {success_count}/{len(gif_files)} thumbnails generated")

if __name__ == "__main__":
    main()

File: mobile/scripts/generate-abdominal-thumbnails.js
#!/usr/bin/env node

/**
 * Generate clean, static thumbnail images from abdominal exercise GIFs.
 * Uses ffmpeg to extract high-quality frames from GIFs.
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const BASE_DIR = path.join(__dirname, '..');
const GIF_DIR = path.join(BASE_DIR, 'assets', 'exercise-gifs', 'abdominals');
const THUMB_DIR = path.join(BASE_DIR, 'assets', 'exercise-thumbnails', 'abdominals');

// Ensure thumbnail directory exists
if (!fs.existsSync(THUMB_DIR)) {
    fs.mkdirSync(THUMB_DIR, { recursive: true });
}

// Get all GIF files
const gifFiles = fs.readdirSync(GIF_DIR).filter(f => f.endsWith('.gif'));

console.log(`Found ${gifFiles.length} GIF files to process\n`);

let successCount = 0;

gifFiles.forEach(gifFile => {
    const gifPath = path.join(GIF_DIR, gifFile);
    const thumbName = gifFile.replace('.gif', '.jpg');
    const thumbPath = path.join(THUMB_DIR, thumbName);
    
    console.log(`Processing: ${gifFile} -> ${thumbName}`);
    
    try {
        // Use ffmpeg to extract a frame from the middle of the GIF
        // -ss 00:00:00.5 seeks to 0.5 seconds (middle of most exercise GIFs)
        // -frames:v 1 extracts just one frame
        // -q:v 2 sets high quality (lower number = higher quality, 2 is very good)
        execSync(
            `ffmpeg -i "${gifPath}" -ss 00:00:00.5 -frames:v 1 -q:v 2 "${thumbPath}" -y 2>/dev/null`,
            { stdio: 'pipe' }
        );
        
        // Verify the file was created
        if (fs.existsSync(thumbPath)) {
            const stats = fs.statSync(thumbPath);
            console.log(`  ✓ Created: ${thumbName} (${Math.round(stats.size / 1024)}KB)`);
            successCount++;
        } else {
            console.log(`  ✗ Failed to create: ${thumbName}`);
        }
    } catch (error) {
        console.log(`  ✗ Error processing ${gifFile}: ${error.message}`);
    }
});

console.log(`\n✨ Completed: ${successCount}/${gifFiles.length} thumbnails generated`);